### PR TITLE
New configuration system of communicator inspired by links

### PR DESCRIPTION
**Pull request by kuenishi**
With this change, communicator configuration can be changed after creation.

### CODE REVIEW

**Comment by kuenishi** [comment]
TODO: make `batched_copy` default within this work.

**Comment by kuenishi** [comment]
Will rebase onto https://github.com/chainer/chainer/pull/7877 once it's merged.

**Comment by kuenishi** [comment]
Rebased to the master that includes #7877 and added enhancements. Ready for review

File: tests/chainermn_tests/communicator_tests/test_communicator.py [hunk 1]
```diff
@@ -899,3 +908,14 @@ def test_send_recv_obj(self):
         self.check_send_recv_obj(5, tag=5, use_any_recv=False, use_status=True)
 
         self.teardown()
+
+    def test_config(self):
```
Comment (shu65) [inline tests/chainermn_tests/communicator_tests/test_communicator.py]: How about adding a test in which `with' statement is used?
`with` statement can be used in the new config system.

**Comment by kuenishi** [comment]
Updated~

**Comment by shu65** [comment]
I'm performing unittests for chainermn

**Comment by shu65** [comment]
I got the following error:
```
tests/chainermn_tests/communicator_tests/test_communicator.py::test_mixed_dtype_communicator_gpu[param7] ESC[31mFAILEDESC[0m
param = {'allreduce_grad_dtype': None,
 'batched_copy': True,
 'communicator_class': <class 'chainermn.communicators.pure_nccl...>,
 'global_dtype': <class 'numpy.float32'>,
 'gpu': False,
 'model_dtype': None,
 'multi_node': True,
 'nccl1': False}
model = <test_communicator.ExampleMixedModel object at 0x7fe8d14f1828>
use_gpu = True

ESC[1m    def check_multi_node_mean_grad_mixed_dtype(param, model, use_gpu):ESC[0m
ESC[1m        # Checks the actual allreduce communication is performedESC[0m
ESC[1m        # in the correct data type (FP16 or FP32)ESC[0m
ESC[1m        comm_class = param.communicator_classESC[0m
ESC[1m    ESC[0m
ESC[1m        if not param.multi_node:ESC[0m
ESC[1m            ranks = _communication_utility.init_ranks(mpi_comm)ESC[0m
ESC[1m            inter_size = ranks[4]ESC[0m
ESC[1m            if inter_size > 1:ESC[0m
ESC[1m                pytest.skip('This test is for single node only')ESC[0m
ESC[1m    ESC[0m
ESC[1m        communicator = comm_class(mpi_comm)ESC[0m
ESC[1m        communicator.set_config('batched_copy', param.batched_copy)ESC[0m
ESC[1m    ESC[0m
ESC[1m        if comm_class is PureNcclCommunicator:ESC[0m
ESC[1m            communicator.set_config('allreduce_grad_dtype',ESC[0m
ESC[1m                                    param.allreduce_grad_dtype)ESC[0m
ESC[1m            value = communicator.get_config('allreduce_grad_dtype')ESC[0m
ESC[1m            assert param.allreduce_grad_dtype == valueESC[0m
ESC[1m    ESC[0m
ESC[1m        mpi_comm.barrier()ESC[0m
ESC[1m    ESC[0m
ESC[1m        # answer type: see the document of `create_communicator`ESC[0m
ESC[1m        global_dtype = param.global_dtypeESC[0m
ESC[1m        allreduce_dtype = param.allreduce_grad_dtypeESC[0m
ESC[1m    ESC[0m
ESC[1m        # assert test configuration.ESC[0m
ESC[1m        assert chainer.get_dtype() == global_dtypeESC[0m
ESC[1m    ESC[0m
ESC[1m        answer_dtype = NoneESC[0m
ESC[1m        if allreduce_dtype == np.float16:ESC[0m
ESC[1m            answer_dtype = np.float16ESC[0m
ESC[1m        elif allreduce_dtype == np.float32:ESC[0m
ESC[1m            answer_dtype = np.float32ESC[0m
ESC[1m        else:ESC[0m
ESC[1m            if global_dtype == np.float32:ESC[0m
ESC[1m                answer_dtype = np.float32ESC[0m
ESC[1m            else:ESC[0m
ESC[1m                answer_dtype = np.float16ESC[0m
ESC[1m    ESC[0m
ESC[1m        if use_gpu:ESC[0m
ESC[1m            model.to_gpu()ESC[0m
ESC[1m    ESC[0m
ESC[1m        model.a.W.grad[:] = communicator.rankESC[0m
ESC[1m        model.b.W.grad[:] = communicator.rank + 1ESC[0m
ESC[1m        model.c.b.grad[:] = communicator.rank + 2ESC[0m
ESC[1m    ESC[0m
ESC[1m        if isinstance(communicator, PureNcclCommunicator):ESC[0m
ESC[1m            communicator._init_comms()ESC[0m
ESC[1m            with mock.patch.object(communicator, 'nccl_comm',ESC[0m
ESC[1m                                   wraps=communicator.nccl_comm) as mc:ESC[0m
ESC[1m                answer_dtype = _communication_utility._get_nccl_type_id(ESC[0m
ESC[1m                    answer_dtype)ESC[0m
ESC[1m    ESC[0m
ESC[1m                communicator.multi_node_mean_grad(model)ESC[0m
ESC[1m    ESC[0m
ESC[1m                # dtype that was used in the actual communication,ESC[0m
ESC[1m                # which is nccl_comm.allReduceESC[0m
ESC[1m                call_args = mc.allReduce.call_args[0]ESC[0m
ESC[1m                actual_dtype = call_args[3]ESC[0m
ESC[1m>               assert answer_dtype == actual_dtypeESC[0m
ESC[1mESC[31mE               assert 7 == 8ESC[0m
ESC[1mESC[31mE                 -7ESC[0m
ESC[1mESC[31mE                 +8ESC[0m

ESC[1mESC[31mtests/chainermn_tests/communicator_tests/test_communicator.pyESC[0m:405: AssertionError
```

**Comment by kuenishi** [comment]
I have fixed the issue.

**Comment by kuenishi** [comment]
Test failure in Travis-CI is not related to this change.

**Comment by shu65** [comment]
Jenkins, test this please

**Comment by shu65** [comment]
I'm performing tests for ChainerMN.

**Comment by shu65** [comment]
Jenkins, test this please

**Comment by shu65** [comment]
Jenkins, test this please

**Comment by kuenishi** [comment]
Test failure both in Jenkins and Travis are not related to this change IMO.

**Pull request merged**